src/huggingface_hub/inference/_mcp/mcp_client.py

```diff
@@ -0,0 +1,236 @@
+import json
+import logging
+import warnings
+from contextlib import AsyncExitStack
+from pathlib import Path
+from typing import TYPE_CHECKING, AsyncIterable, Dict, List, Optional, Union
+
+from typing_extensions import TypeAlias
+
+from ...utils._runtime import get_hf_hub_version
+from .._generated._async_client import AsyncInferenceClient
+from .._generated.types import (
+    ChatCompletionInputMessage,
+    ChatCompletionInputTool,
+    ChatCompletionStreamOutput,
+    ChatCompletionStreamOutputDeltaToolCall,
+)
+from .._providers import PROVIDER_OR_POLICY_T
+from .utils import format_result
+
+
+if TYPE_CHECKING:
+    from mcp import ClientSession
+
+logger = logging.getLogger(__name__)
+
+# Type alias for tool names
+ToolName: TypeAlias = str
+
+
+class MCPClient:
+    def __init__(
+        self,
+        *,
+        model: str,
+        provider: Optional[PROVIDER_OR_POLICY_T] = None,
+        api_key: Optional[str] = None,
+    ):
+        warnings.warn(
+            "'MCPClient' is experimental and might be subject to breaking changes in the future without prior notice.",
+            UserWarning,
+        )
+
+        # Initialize MCP sessions as a dictionary of ClientSession objects
+        self.sessions: Dict[ToolName, "ClientSession"] = {}
+        self.exit_stack = AsyncExitStack()
+        self.available_tools: List[ChatCompletionInputTool] = []
+
+        # Initialize the AsyncInferenceClient
+        self.client = AsyncInferenceClient(model=model, provider=provider, api_key=api_key)
+
+    async def __aenter__(self):
+        """Enter the context manager"""
+        await self.client.__aenter__()
```

Comment:
just for the sake of ressource leakage, maybe we should enter `self.exist_stack` either here or we can do:
```python
async with elf.exit_stack:
      # call self.exit_stack.enter_async_context(..)
```
```suggestion
        await self.client.__aenter__()
        await self.exit_stack.__aenter__()
```